### PR TITLE
feat(ai-engine): support more document formats in ingestion (DOCX, CSV, TXT, Markdown, HTML)

### PR DESCRIPTION
**Pull request by chaitanya21kumar**
## What this adds

More input formats for Module A.1 ingestion — **DOCX, CSV, TXT, Markdown, HTML** — alongside the existing PDF and PPTX. Each parses into the same `ParsedDocument` contract, so `POST /ingest` and `POST /summarize/file` accept them with no other changes.

## Note on the diff (builds on #56)

These parsers register in the shared ingestion service introduced by #56 (Module A.2), so this branch is stacked on #56. **Until #56 merges, this PR's diff also includes the A.2 commits; it narrows to just the format work once #56 is merged.** The files to review here are:

- `app/ingestion/schema.py` — additive: new `format` values + two `Page.kind` values (`document`, `sheet`)
- `app/ingestion/textio.py` + `{txt,csv,md,html,docx}_parser.py` — the five parsers
- `app/ingestion/service.py` — register the parsers by extension
- `tests/test_format_parsers.py` — coverage

## Formats and how they map to the block schema

- **DOCX** (python-docx) — headings, paragraphs, lists, tables, images
- **CSV** (standard library) — one table on a `sheet` page; delimiter sniffed, large files capped
- **TXT** (standard library) — paragraphs and simple lists
- **Markdown** (markdown-it-py) — headings, lists, GFM tables, code, paragraphs; inline markup flattened
- **HTML** (beautifulsoup4) — headings, paragraphs, lists, tables, images; script/style stripped

Flow formats (DOCX, TXT, Markdown, HTML) are one `document` page; CSV is one `sheet`. The shape is unchanged, so the summarisation layer and future modules work on these formats unchanged.

## How it was tested

- **47 unit tests** (13 new): per-parser structure (headings, lists, tables, flattened markup, stripped scripts, CRLF handling, empty-cell and image-in-paragraph edge cases) plus endpoint wiring; the two old "unsupported type" tests updated since `.txt` is now supported.
- **Live**: all five formats ingest through `POST /ingest`, and a DOCX summarises end to end through `POST /summarize/file` — summarisation works on the new formats with no changes.

Refs #7

### CODE REVIEW

File: ai-engine/app/ingestion/docx_parser.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    if name.startswith("Heading "):
+        try:
+            return min(int(name.split()[1]), 6)
+        except ValueError:
+            return 1
+    if name == "Title":
+        return 1
+    if name == "Subtitle":
+        return 2
+    return None
```
[the remark this answers is not in the document]
Comment (chaitanya21kumar) [reply ai-engine/app/ingestion/docx_parser.py:50]: Good catch - fixed. _heading_level now checks the style name has a numeric part before indexing, so a malformed "Heading " style returns level 1 instead of raising an IndexError. Pushed in 70ee9c4.

File: ai-engine/app/ingestion/md_parser.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        elif t == "tr_close":
+            if current:
+                rows.append(current)
+            current = None
+        elif t == "inline" and current is not None:
+            current.append(_inline_plain(tokens[i]))
+        elif t == "table_close":
+            return rows, i + 1
+        i += 1
+    return rows, i
```
[the remark this answers is not in the document]
Comment (chaitanya21kumar) [reply ai-engine/app/ingestion/md_parser.py:112]: Fixed - the table collector now tracks each th/td open and close and appends one value per cell (an empty string when the cell is blank), so columns stay aligned. Added a test with an empty cell. 70ee9c4.

File: ai-engine/app/ingestion/html_parser.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            if text:
+                blocks.append(Block(kind=BlockKind.heading, text=text, level=int(name[1])))
+        elif name == "p":
+            text = child.get_text(" ", strip=True)
+            if text:
+                blocks.append(Block(kind=BlockKind.paragraph, text=text))
+            else:
+                for img in child.find_all("img"):
+                    counter[0] += 1
+                    blocks.append(_image_block(img, counter[0]))
```
[the remark this answers is not in the document]
Comment (chaitanya21kumar) [reply ai-engine/app/ingestion/html_parser.py]: Fixed - images inside a paragraph are now captured whether or not the paragraph also has text. Added a test. 70ee9c4.

File: ai-engine/app/summarization/llm_client.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        "stream": False,
+        "response_format": {"type": "json_object"},
+    }
+    headers = {"Authorization": f"Bearer {api_key}"}
+
+    for attempt in range(max_retries + 1):
+        try:
+            response = await client.post(
+                f"{base_url}/chat/completions", json=payload, headers=headers
+            )
```
[the remark this answers is not in the document]
Comment (chaitanya21kumar) [reply ai-engine/app/summarization/llm_client.py:80]: Fixed at the root - llm_base_url is normalised in config to strip any trailing slash, so the chat-completions URL and the /ready probe are always well-formed. Added a test. 70ee9c4.

**Comment by chaitanya21kumar** [comment]
Proof — 47 tests passing, every supported format ingesting, and a DOCX summarised live:
pr57-tests-passing
pr57-docx-summarised
pr57-all-formats-ingest

**Comment by siddhishinde0723** [comment]
🤖 **AI Code Review Summary**

### 📝 Overview
The codebase is clean, well-architected, and follows modern FastAPI/Pydantic patterns. The modular design of the ingestion parsers and the clear abstraction of the LLM client are excellent for maintainability. The most significant risks concern potential resource exhaustion with large files and the need for stricter timeout management in the pipeline. --- The PR demonstrates high-quality engineering, particularly in the separation of concerns between ingestion and summarization and the inclusion of comprehensive unit testing. The architecture is well-aligned with the goal of being model-agnostic, though there are minor issues regarding robustness and consistency in the newly added ingestion logic.

### 🚩 Top Findings

#### [MAJOR] Potential Resource Exhaustion in `csv_parser.py`
- **Location**: `ai-engine/app/ingestion/csv_parser.py:35`
- **Issue**: The code reads the entire file into memory via `read_text(path)` before passing it to `csv.reader`. While `_MAX_ROWS` limits the *output* blocks, a multi-gigabyte CSV will crash the container during the initial `read_text` call.

#### [MAJOR] Unchecked LLM Pipeline Latency
- **Location**: `ai-engine/app/summarization/pipeline.py:157`
- **Issue**: `generate_insights` performs three sequential network calls to the LLM gateway. If each takes close to the `llm_request_timeout` (120s), a single request could hang for 6 minutes, potentially leading to gateway-level timeouts (e.g., Nginx or Load Balancer).

#### [MAJOR] Lack of Input Sanitization in `parse_html`
- **Location**: `ai-engine/app/ingestion/html_parser.py` (Implied by `test_html_structure_and_script_stripped`)
- **Issue**: While the test ensures `<script>` tags are removed, raw HTML ingestion is prone to other XSS vectors (e.g., `onerror` handlers, `<iframe>`, `<a>` with `javascript:` links). As this data is likely passed to an LLM, it poses a prompt injection risk if the source content is untrusted.

**Comment by chaitanya21kumar** [comment]
Thank you, ma'am. On the three findings:

1. **Large CSV / memory:** fixed at the boundary rather than only in the CSV parser — uploads are streamed to disk with a size ceiling (`AI_ENGINE_MAX_UPLOAD_BYTES`, default 25 MiB) and anything over it is rejected with a 413 before it's fully read, so memory is bounded for every format, not just CSV. — `a66f52c`
2. **Sequential LLM latency:** kept sequential on purpose (also discussed on #56). Dev runs against a hosted free tier and prod self-hosts a single model; firing the three calls concurrently bursts the per-minute rate limit (we already hit 429s on larger docs) and a single local model serialises them anyway, so `gather` wouldn't reduce wall-clock time there. Retry/backoff handles transient limits. Worth revisiting with `asyncio.gather` on a paid or scaled deployment where concurrency clearly wins.
3. **HTML sanitisation:** the parser only emits extracted text (`get_text` plus an image's alt/size), never attributes or raw markup — so `onerror=`, `javascript:` hrefs and `<iframe>` never reach the output and nothing is rendered, so browser-XSS doesn't apply. Instruction-injection via the visible text is real but common to every format (a PDF can say "ignore instructions"), and the system prompt already handles that by treating the source as data. As defence in depth I also drop iframe/svg/object/embed/template now. — `5f6892a`

51 tests green and the SonarCloud quality gate is passing. Thanks again, ma'am.

**Comment by chaitanya21kumar** [comment]
Thanks for the review — all points addressed:

- **Docstrings that still said PDF/PPTX only** (ingestion router + `ingest()`, `parse_upload`, and `/summarize/file`): reworded to "an uploaded document / the supported formats", since parsing is delegated to the parser registry. — `fefed22`
- **README** (Module A.2 line + `/ingest` and `/summarize/file` bullets): now list the full format set — PDF, PPTX, DOCX, CSV, TXT, Markdown, HTML — and the OpenAI-compatible gateway instead of Ollama. — `8196c85`, `fefed22`
- **Package docstring claiming Ollama-only:** reworded to the provider-agnostic gateway. — `fefed22`
- **Fixture leaking the fake client:** it now closes the `AsyncClient`(s) on teardown. — `cef6f74`

51 tests green.
